### PR TITLE
add load balancers and volumes finilizers for delete cluster

### LINKED ISSUE #2602: Setup `kubermatic.io/cleanup-in-cluster-pv` and/or `kubermatic.io/cleanup-in-cluster-lb` finalizers in backend for delete cluster operation
**User Story**
Depending on what the user selected the `kubermatic.io/cleanup-in-cluster-pv` and/or `kubermatic.io/cleanup-in-cluster-lb` finalizers must be added to the cluster object prior to the DELETE call

**Acceptance criteria**
DELETE call checks header params to get information about mentioned finalizers.
If some finalizer is added then cluster must be updated to reflect added finilizer and then deleted.

### PR DESCRIPTION
**Pull request by zreigz**
**What this PR does / why we need it**: it is possible for a user to select if PVC's/PV's and/or LB's should be cleaned up when deleting the cluster.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2602

**Release note**:
```release-note
It is now possible for a user to select whether PVCs/PVs and/or LBs should be cleaned up when deleting the cluster.
```

### CODE REVIEW

File: api/pkg/handler/v1/common/request.go [hunk 1]
```diff
@@ -62,6 +63,48 @@ type GetClusterReq struct {
 	ClusterID string `json:"cluster_id"`
 }
 
+type DeleteClusterReq struct {
+	DCReq
+	// in: path
+	ClusterID string `json:"cluster_id"`
+	// DeleteVolumes if true all cluster PV's and PVC's will be deleted from cluster
+	DeleteVolumes bool
+	// DeleteVolumes if true all load balancers will be deleted from cluster
```
Comment (p0lyn0mial) [inline api/pkg/handler/v1/common/request.go]: change to `DeleteLoadBalancers`

File: api/pkg/handler/v1/common/request.go [hunk 2]
```diff
@@ -62,6 +63,48 @@ type GetClusterReq struct {
 	ClusterID string `json:"cluster_id"`
 }
 
+type DeleteClusterReq struct {
```
Comment (p0lyn0mial) [inline api/pkg/handler/v1/common/request.go]: could be moved to `cluster.go`, `common` pkg should be reserved for types that are used by many handlers.

**Review by p0lyn0mial — COMMENTED** [review]
@zreigz code looks good, could you add a unit test ?

File: api/pkg/handler/v1/common/request.go (hunk 2 above)
Comment (zreigz) [reply api/pkg/handler/v1/common/request.go]: unfortunately we have cyclic package dependency problem in this case:
```
import cycle not allowed
package github.com/kubermatic/kubermatic/api/cmd/kubermatic-api
	imports github.com/kubermatic/kubermatic/api/pkg/handler
	imports github.com/kubermatic/kubermatic/api/pkg/handler/v1/common
	imports github.com/kubermatic/kubermatic/api/pkg/handler
can't load package: import cycle not allowed
package github.com/kubermatic/kubermatic/api/cmd/kubermatic-api
	imports github.com/kubermatic/kubermatic/api/pkg/handler
	imports github.com/kubermatic/kubermatic/api/pkg/handler/v1/common
	imports github.com/kubermatic/kubermatic/api/pkg/handler
```

**Comment by zreigz** [comment]
@p0lyn0mial yes sure, I will add unit tests, so please hold on until a new commit

File: api/pkg/handler/v1/common/request.go (hunk 2 above)
Comment (p0lyn0mial) [reply api/pkg/handler/v1/common/request.go]: do we https://github.com/kubermatic/kubermatic/pull/2606 ?

Comment (zreigz) [reply api/pkg/handler/v1/common/request.go]: I've just moved only `DeleteClusterReq` structure. PR `#2606` works fine :)

File: api/pkg/handler/cluster.go [hunk 3]
```diff
@@ -671,3 +692,45 @@ func revokeClusterAdminToken(projectProvider provider.ProjectProvider) endpoint.
 		return nil, common.KubernetesErrorToHTTPError(err)
 	}
 }
+
+type DeleteClusterReq struct {
+	common.DCReq
```
Comment (p0lyn0mial) [inline api/pkg/handler/cluster.go]: I think that `common.DCReq` and `ClusterID` could be replaced by `common.GetClusterReq`, wdyt ?

File: api/pkg/handler/v1/common/request.go (hunk 2 above)
Comment (p0lyn0mial) [reply api/pkg/handler/v1/common/request.go]: Ok, thx.

File: api/pkg/handler/v1/common/request.go (hunk 1 above)
Comment (zreigz) [reply api/pkg/handler/v1/common/request.go]: doen

File: api/pkg/handler/cluster.go (hunk 3 above)
Comment (zreigz) [reply api/pkg/handler/cluster.go]: done

File: api/pkg/handler/cluster.go [hunk 4]
```diff
@@ -166,6 +169,24 @@ func deleteCluster(sshKeyProvider provider.SSHKeyProvider, projectProvider provi
...
+			}
+			if req.DeleteLoadBalancers {
+				existingCluster.Finalizers = kuberneteshelper.AddFinalizer(existingCluster.Finalizers, finalizer.InClusterLBCleanupFinalizer)
+			}
+			if req.DeleteVolumes {
+				existingCluster.Finalizers = kuberneteshelper.AddFinalizer(existingCluster.Finalizers, finalizer.InClusterPVCleanupFinalizer)
+			}
+
+			_, err = clusterProvider.Update(userInfo, existingCluster)
+			if err != nil {
```
Comment (mrIncompetent) [inline api/pkg/handler/cluster.go]: If there's nothing except for `err` we consume from the function we could do it within the `if` statement:
```go
if _, err = clusterProvider.Update(userInfo, existingCluster); err != nil {
    return nil, common.KubernetesErrorToHTTPError(err)
}
```

**Comment by mrIncompetent** [comment]
/approve
leaving the lgtm up to @p0lyn0mial

File: api/pkg/handler/cluster.go (hunk 4 above)
Comment (zreigz) [reply api/pkg/handler/cluster.go]: thanks for the tip, I will fix this

**Comment by zreigz** [comment]
@p0lyn0mial PTAL, unit tests added

File: api/pkg/handler/cluster_test.go [hunk 5]
```diff
@@ -21,6 +21,90 @@ import (
...
+			}
+			compareWithResult(t, res, tc.ExpectedResponse)
+
+			validatedActions := 0
+			for _, action := range kubermaticClient.Actions() {
+				if action.Matches("update", "clusters") {
+					validatedActions = validatedActions + 1
+				}
+			}
+			if validatedActions != tc.ExpectedUpdates {
```
Comment (p0lyn0mial) [inline api/pkg/handler/cluster_test.go:137]: I think we could get updated object and examine if it has finalizers attached, for example https://github.com/kubermatic/kubermatic/blob/master/api/pkg/controller/rbac/sync_resource_test.go#L567

File: api/pkg/handler/cluster_test.go [hunk 6]
```diff
@@ -21,6 +21,90 @@ import (
...
+			ExpectedResponse: `{}`,
+			HTTPStatus:       http.StatusOK,
+			ProjectToSync:    genDefaultProject().Name,
+			ExistingKubermaticObjs: genDefaultKubermaticObjects(
+				// add a cluster
+				genCluster("clusterAbcID", "clusterAbc", genDefaultProject().Name, time.Date(2013, 02, 03, 19, 54, 0, 0, time.UTC)),
+			),
+			ClusterToSync:   "clusterAbcID",
+			ExistingAPIUser: genDefaultAPIUser(),
+			HeaderParams:    map[string]string{"DeleteVolumes": "true", "DeleteLoadBalancers": "true"},
```
Comment (p0lyn0mial) [inline api/pkg/handler/cluster_test.go]: we could easily (copy&paste)  add additional test case where `DeleteVolumes is set to false` and `DeleteLoadBalancers is set to true`, wdyt ?

File: api/pkg/handler/cluster_test.go [hunk 7]
```diff
@@ -21,6 +21,90 @@ import (
...
+			ExpectedUpdates: 1,
+		},
+		{
+			Name:             "scenario 1: tests deletion of a cluster without finalizers",
```
Comment (p0lyn0mial) [inline api/pkg/handler/cluster_test.go]: nit: scenario 2

File: api/pkg/handler/cluster_test.go (hunk 5 above)
Comment (zreigz) [reply api/pkg/handler/cluster_test.go:137]: done

File: api/pkg/handler/cluster_test.go (hunk 6 above)
Comment (zreigz) [reply api/pkg/handler/cluster_test.go]: I like it, done

File: api/pkg/handler/cluster_test.go (hunk 7 above)
Comment (zreigz) [reply api/pkg/handler/cluster_test.go]: done

**Comment by zreigz** [comment]
@p0lyn0mial PTAL

File: api/pkg/handler/cluster_test.go [hunk 8]
```diff
@@ -21,6 +23,125 @@ import (
...
+
+					cluster := updateaction.GetObject().(*kubermaticv1.Cluster)
+					finalizers := cluster.GetFinalizers()
+					if len(finalizers) != len(tc.ExpectedFinalizers) {
+						t.Fatalf("not all finalizers were validated, expected to validate %d but validated only %d", len(tc.ExpectedFinalizers), len(finalizers))
+					}
+
+					sort.Strings(finalizers)
+					sort.Strings(tc.ExpectedFinalizers)
+					if !reflect.DeepEqual(finalizers, tc.ExpectedFinalizers) {
```
Comment (p0lyn0mial) [inline api/pkg/handler/cluster_test.go]: Although in this case `reflect.DeepEqual` is perfectly valid we decided no to use it, see https://github.com/kubermatic/kubermatic/blob/master/api/hack/verify-forbidden-functions.sh#L8

Comment (zreigz) [reply api/pkg/handler/cluster_test.go]: done

**Comment by zreigz** [comment]
@p0lyn0mial PTAL

**Comment by zreigz** [comment]
@mrIncompetent @p0lyn0mial thank you for review

**Comment by p0lyn0mial** [comment]
@zreigz np, thank you for the pull :)

**Pull request merged**